packages/core/lib/patchers/aws3_p.ts

```diff
@@ -0,0 +1,179 @@
+import type {
+  MetadataBearer,
+  ResponseMetadata,
+  Command,
+  Client,
+} from '@aws-sdk/types';
+
+import type { RegionInputConfig } from '@aws-sdk/config-resolver';
+
+import { isThrottlingError } from '@aws-sdk/service-error-classification';
+
+import ServiceSegment from '../segments/attributes/aws';
+
+import { stringify } from 'querystring';
+
+import {
+  Subsegment,
+  Segment,
+  getNamespace,
+  resolveSegment,
+  setSegment,
+} from '..';
+
+
+var contextUtils = require('../context_utils');
+
+var logger = require('../logger');
+
+import { getCauseTypeFromHttpStatus } from '../utils';
+
+type HttpResponse = { response: { status: number } };
+
+async function buildAttributesFromMetadata(
+  client: any,
+  command: any,
+  metadata: ResponseMetadata,
+): Promise<[ServiceSegment, HttpResponse]> {
+  const { httpHeaders, httpStatusCode: statusCode, retries } = metadata;
+  const serviceIdentifier = client.config.signingName as string;
+
+  let operation: string = command.constructor.name.slice(0, -7);
+  operation = operation.charAt(0).toLowerCase() + operation.slice(1);
+
+  const aws = new ServiceSegment(
+    {
+      extendedRequestId: httpHeaders && httpHeaders['x-amz-id-2'],
+      requestId: httpHeaders && httpHeaders['x-amz-request-id'],
+      retryCount: retries,
+      request: {
+        operation,
+        httpRequest: {
+          region: await client.config.region(),
+          statusCode,
+        },
+        params: command.input,
+      },
+    },
+    serviceIdentifier,
+  );
+
+  const http = { response: { status: statusCode || 0 } };
+  return [aws, http];
+}
+
+function addFlags(http: HttpResponse, subsegment: Subsegment, err?: Error): void {
+  if (err && isThrottlingError(err)) {
+    subsegment.addThrottleFlag();
+  } else if (http.response?.status === 429) {
+    subsegment.addThrottleFlag();
+  }
+
+  const cause = getCauseTypeFromHttpStatus(http.response.status);
+  if (cause === 'fault') {
+    subsegment.addFaultFlag();
+  } else if (cause === 'error') {
+    subsegment.addErrorFlag();
+  }
+}
+
+type DefaultConfiguration = RegionInputConfig & {
+  signingName: string;
+};
+
+export function captureAWSClient<
+  Input extends object,
+  Output extends MetadataBearer,
+  Configuration extends DefaultConfiguration
+>(client: Client<Input, Output, Configuration>): Client<Input, Output, Configuration> {
+  // create local copy so that we can later call it
+  const send = client.send;
+
+  const serviceIdentifier = client.config.signingName;
+
+  client.send = async (command: any): Promise<Output> => {
```

Comment:
Is there any way we could also use the `middlewareStack` for this instead of redefining the `send` operation? Totally ok if not, I'd just rather use the first-class features of the new AWS SDK where possible :)

Reply:
I tried the middleware stack but was unable to capture any exceptions. The middleware stack only gave me access to the raw response so that would force us to build the exception in the patcher. I'd love for someone from the v3 team to chime in and propose a better solution.

Reply:
Not sure if he'll respond but: @trivikr do you know of any better way to capture exceptions from a AWS SDK V3 request using the middleware stack? Otherwise we have to use a workaround by surrounding the requests with a `try/catch` like we are here.